File: snippets/screens/compose/biometric_auth_screen/2/biometric_auth_screen.kt
package com.example.auth.screens

import androidx.biometric.BiometricManager
import androidx.biometric.BiometricPrompt
import androidx.compose.animation.core.*
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.selection.selectable
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.CheckCircle
import androidx.compose.material.icons.filled.Error
import androidx.compose.material.icons.filled.Fingerprint
import androidx.compose.material.icons.filled.Security
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.scale
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.core.content.ContextCompat
import androidx.fragment.app.FragmentActivity
import androidx.lifecycle.viewmodel.compose.viewModel
import kotlinx.coroutines.launch

enum class BiometricType(
    val displayName: String,
    val icon: ImageVector,
    val description: String
) {
    FINGERPRINT("Fingerprint", Icons.Default.Fingerprint, "Use your fingerprint to authenticate"),
    FACE("Face Recognition", Icons.Default.Security, "Use your face to authenticate"),
    IRIS("Iris Recognition", Icons.Default.Security, "Use your eyes to authenticate"),
    NONE("Biometric", Icons.Default.Security, "Use biometric authentication")
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun BiometricAuthScreen(
    isSetup: Boolean = false,
    viewModel: BiometricAuthViewModel = viewModel(),
    onNavigateBack: () -> Unit = {},
    onSetupComplete: () -> Unit = {},
    onAuthenticationSuccess: () -> Unit = {}
) {
    val context = LocalContext.current
    val scrollState = rememberScrollState()
    val snackbarHostState = remember { SnackbarHostState() }
    val coroutineScope = rememberCoroutineScope()

    LaunchedEffect(Unit) {
        viewModel.checkBiometricAvailability(context)
        viewModel.isSetup = isSetup
    }

    LaunchedEffect(viewModel.authenticationSuccess) {
        if (viewModel.authenticationSuccess) {
            if (isSetup) {
                onSetupComplete()
            } else {
                coroutineScope.launch {
                    snackbarHostState.showSnackbar("Biometric authentication successful!")
                }
                onAuthenticationSuccess()
            }
        }
    }

    Scaffold(
        topBar = {
            if (isSetup) {
                TopAppBar(
                    title = { Text("Setup Biometric Auth") },
                    navigationIcon = {
                        IconButton(onClick = onNavigateBack) {
                            Icon(Icons.Default.Security, contentDescription = "Back")
                        }
                    }
                )
            }
        },
        snackbarHost = { SnackbarHost(snackbarHostState) }
    ) { paddingValues ->
        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(paddingValues)
                .padding(24.dp)
                .verticalScroll(scrollState),
            horizontalAlignment = Alignment.CenterHorizontally
        ) {
            Spacer(modifier = Modifier.height(if (isSetup) 20.dp else 60.dp))

            when {
                viewModel.setupComplete -> SetupCompleteContent(onContinue = onSetupComplete)
                !viewModel.isAvailable -> UnavailableContent(
                    error = viewModel.error,
                    isSetup = isSetup,
                    onCheckAgain = { viewModel.checkBiometricAvailability(context) },
                    onSkip = if (isSetup) onSetupComplete else onNavigateBack
                )
                else -> AuthenticationContent(
                    isSetup = isSetup,
                    biometricType = viewModel.biometricType,
                    availableBiometrics = viewModel.availableBiometrics,
                    selectedBiometric = viewModel.selectedBiometric,
                    onBiometricSelected = viewModel::selectBiometric,
                    isLoading = viewModel.isLoading,
                    error = viewModel.error,
                    onAuthenticate = {
                        if (context is FragmentActivity) {
                            viewModel.authenticateWithBiometric(context)
                        }
                    },
                    onSkip = if (isSetup) onSetupComplete else onNavigateBack
                )
            }

            Spacer(modifier = Modifier.height(40.dp))
        }
    }
}

@Composable
fun AuthenticationContent(
    isSetup: Boolean,
    biometricType: BiometricType,
    availableBiometrics: List<BiometricType>,
    selectedBiometric: BiometricType,
    onBiometricSelected: (BiometricType) -> Unit,
    isLoading: Boolean,
    error: String?,
    onAuthenticate: () -> Unit,
    onSkip: () -> Unit
) {
    // Animated biometric icon
    val infiniteTransition = rememberInfiniteTransition(label = "biometric_animation")
    val scale by infiniteTransition.animateFloat(
        initialValue = 1f,
        targetValue = 1.1f,
        animationSpec = infiniteRepeatable(
            animation = tween(1500, easing = EaseInOut),
            repeatMode = RepeatMode.Reverse
        ),
        label = "scale_animation"
    )

    Icon(
        imageVector = biometricType.icon,
        contentDescription = null,
        modifier = Modifier
            .size(120.dp)
            .scale(scale),
        tint = MaterialTheme.colorScheme.primary
    )

    Spacer(modifier = Modifier.height(40.dp))

    // Header
    Text(
        text = if (isSetup) "Setup ${biometricType.displayName}" else "Biometric Authentication",
        style = MaterialTheme.typography.headlineSmall,
        fontWeight = FontWeight.Bold,
        textAlign = TextAlign.Center
    )

    Spacer(modifier = Modifier.height(12.dp))

    Text(
        text = if (isSetup) {
            "Enable ${biometricType.displayName.lowercase()} authentication for quick and secure access to your account."
        } else {
            biometricType.description
        },
        style = MaterialTheme.typography.bodyLarge,
        color = MaterialTheme.colorScheme.onSurfaceVariant,
        textAlign = TextAlign.Center
    )

    Spacer(modifier = Modifier.height(40.dp))

    // Available biometric methods (if multiple)
    if (availableBiometrics.size > 1) {
        BiometricSelectionCard(
            availableBiometrics = availableBiometrics,
            selectedBiometric = selectedBiometric,
            onBiometricSelected = onBiometricSelected
        )
        Spacer(modifier = Modifier.height(32.dp))
    }

    error?.let {
        ErrorCard(message = it)
        Spacer(modifier = Modifier.height(24.dp))
    }

    // Security info card
    SecurityInfoCard()

    Spacer(modifier = Modifier.height(32.dp))

    // Authenticate button
    Button(
        onClick = onAuthenticate,
        enabled = !isLoading,
        modifier = Modifier
            .fillMaxWidth()
            .height(56.dp)
    ) {
        if (isLoading) {
            CircularProgressIndicator(
                modifier = Modifier.size(20.dp),
                color = MaterialTheme.colorScheme.onPrimary
            )
        } else {
            Text(if (isSetup) "Enable ${biometricType.displayName}" else "Authenticate")
        }
    }

    Spacer(modifier = Modifier.height(16.dp))

    if (isSetup) {
        TextButton(onClick = onSkip) {
            Text("Skip for Now")
        }
    } else {
        OutlinedButton(
            onClick = onSkip,
            modifier = Modifier
                .fillMaxWidth()
                .height(56.dp)
        ) {
            Text("Use Password Instead")
        }
    }
}

@Composable
fun UnavailableContent(
    error: String?,
    isSetup: Boolean,
    onCheckAgain: () -> Unit,
    onSkip: () -> Unit
) {
    // Error icon
    Icon(
        imageVector = Icons.Default.Error,
        contentDescription = null,
        modifier = Modifier.size(80.dp),
        tint = MaterialTheme.colorScheme.error
    )

    Spacer(modifier = Modifier.height(32.dp))

    // Header
    Text(
        text = "Biometric Authentication Unavailable",
        style = MaterialTheme.typography.headlineSmall,
        fontWeight = FontWeight.Bold,
        textAlign = TextAlign.Center
    )

    Spacer(modifier = Modifier.height(12.dp))

    Text(
        text = error ?: "Biometric authentication is not available on this device or no biometrics are enrolled.",
        style = MaterialTheme.typography.bodyLarge,
        color = MaterialTheme.colorScheme.onSurfaceVariant,
        textAlign = TextAlign.Center
    )

    Spacer(modifier = Modifier.height(40.dp))

    // Instructions card
    Card(
        modifier = Modifier.fillMaxWidth(),
        colors = CardDefaults.cardColors(
            containerColor = MaterialTheme.colorScheme.warningContainer
        )
    ) {
        Column(
            modifier = Modifier.padding(16.dp)
        ) {
            Row(
                verticalAlignment = Alignment.CenterVertically
            ) {
                Icon(
                    imageVector = Icons.Default.Error,
                    contentDescription = null,
                    tint = MaterialTheme.colorScheme.onWarningContainer
                )
                Spacer(modifier = Modifier.width(12.dp))
                Text(
                    text = "To use biometric authentication:",
                    style = MaterialTheme.typography.titleSmall,
                    fontWeight = FontWeight.SemiBold,
                    color = MaterialTheme.colorScheme.onWarningContainer
                )
            }

            Spacer(modifier = Modifier.height(12.dp))

            Column(
                verticalArrangement = Arrangement.spacedBy(8.dp)
            ) {
                Text(
                    text = "• Enable fingerprint or face unlock in device settings",
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onWarningContainer
                )
                Text(
                    text = "• Make sure at least one biometric is enrolled",
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onWarningContainer
                )
                Text(
                    text = "• Restart the app after enabling biometrics",
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onWarningContainer
                )
            }
        }
    }

    Spacer(modifier = Modifier.height(32.dp))

    // Check again button
    Button(
        onClick = onCheckAgain,
        modifier = Modifier
            .fillMaxWidth()
            .height(56.dp)
    ) {
        Text("Check Again")
    }

    Spacer(modifier = Modifier.height(16.dp))

    // Skip/alternative button
    OutlinedButton(
        onClick = onSkip,
        modifier = Modifier
            .fillMaxWidth()
            .height(56.dp)
    ) {
        Text(if (isSetup) "Skip Setup" else "Use Password Instead")
    }
}

@Composable
fun SetupCompleteContent(onContinue: () -> Unit) {
    // Success icon
    Box(
        modifier = Modifier
            .size(120.dp)
            .padding(24.dp),
        contentAlignment = Alignment.Center
    ) {
        Card(
            modifier = Modifier.fillMaxSize(),
            colors = CardDefaults.cardColors(
                containerColor = Color.Green.copy(alpha = 0.1f)
            ),
            shape = androidx.compose.foundation.shape.CircleShape
        ) {}
        Icon(
            imageVector = Icons.Default.CheckCircle,
            contentDescription = null,
            modifier = Modifier.size(80.dp),
            tint = Color.Green
        )
    }

    Spacer(modifier = Modifier.height(24.dp))

    // Header
    Text(
        text = "Biometric Authentication Enabled!",
        style = MaterialTheme.typography.headlineSmall,
        fontWeight = FontWeight.Bold,
        textAlign = TextAlign.Center
    )

    Spacer(modifier = Modifier.height(16.dp))

    Text(
        text = "You can now use biometric authentication to quickly and securely access your account.",
        style = MaterialTheme.typography.bodyLarge,
        color = MaterialTheme.colorScheme.onSurfaceVariant,
        textAlign = TextAlign.Center
    )

    Spacer(modifier = Modifier.height(40.dp))

    // Features card
    Card(
        modifier = Modifier.fillMaxWidth(),
        colors = CardDefaults.cardColors(
            containerColor = MaterialTheme.colorScheme.primaryContainer
        )
    ) {
        Column(
            modifier = Modifier.padding(16.dp)
        ) {
            Row(
                verticalAlignment = Alignment.CenterVertically
            ) {
                Icon(
                    imageVector = Icons.Default.CheckCircle,
                    contentDescription = null,
                    tint = MaterialTheme.colorScheme.onPrimaryContainer
                )
                Spacer(modifier = Modifier.width(12.dp))
                Text(
                    text = "Biometric Authentication Features:",
                    style = MaterialTheme.typography.titleSmall,
                    fontWeight = FontWeight.SemiBold,
                    color = MaterialTheme.colorScheme.onPrimaryContainer
                )
            }

            Spacer(modifier = Modifier.height(12.dp))

            Column(
                verticalArrangement = Arrangement.spacedBy(8.dp)
            ) {
                Text(
                    text = "• Quick and secure login",
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onPrimaryContainer
                )
                Text(
                    text = "• No need to remember passwords",
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onPrimaryContainer
                )
                Text(
                    text = "• Your biometric data stays on your device",
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onPrimaryContainer
                )
                Text(
                    text = "• Can be disabled anytime in settings",
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onPrimaryContainer
                )
            }
        }
    }

    Spacer(modifier = Modifier.height(32.dp))

    // Continue button
    Button(
        onClick = onContinue,
        modifier = Modifier
            .fillMaxWidth()
            .height(56.dp)
    ) {
        Text("Continue to App")
    }

    Spacer(modifier = Modifier.height(16.dp))

    // Test button
    OutlinedButton(
        onClick = { /* Test biometric */ },
        modifier = Modifier
            .fillMaxWidth()
            .height(56.dp)
    ) {
        Text("Test Biometric Login")
    }
}

@Composable
fun BiometricSelectionCard(
    availableBiometrics: List<BiometricType>,
    selectedBiometric: BiometricType,
    onBiometricSelected: (BiometricType) -> Unit
) {
    Card(
        modifier = Modifier.fillMaxWidth()
    ) {
        Column(
            modifier = Modifier.padding(16.dp)
        ) {
            Text(
                text = "Available Authentication Methods:",
                style = MaterialTheme.typography.titleMedium,
                fontWeight = FontWeight.SemiBold
            )

            Spacer(modifier = Modifier.height(12.dp))

            availableBiometrics.forEach { biometric ->
                Row(
                    modifier = Modifier
                        .fillMaxWidth()
                        .selectable(
                            selected = selectedBiometric == biometric,
                            onClick = { onBiometricSelected(biometric) }
                        )
                        .padding(vertical = 4.dp),
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    RadioButton(
                        selected = selectedBiometric == biometric,
                        onClick = { onBiometricSelected(biometric) }
                    )
                    Spacer(modifier = Modifier.width(12.dp))
                    Icon(
                        imageVector = biometric.icon,
                        contentDescription = null,
                        modifier = Modifier.size(24.dp)
                    )
                    Spacer(modifier = Modifier.width(12.dp))
                    Text(
                        text = biometric.displayName,
                        style = MaterialTheme.typography.bodyLarge
                    )
                }
            }
        }
    }
}

@Composable
fun SecurityInfoCard() {
    Card(
        modifier = Modifier.fillMaxWidth(),
        colors = CardDefaults.cardColors(
            containerColor = MaterialTheme.colorScheme.surfaceVariant
        )
    ) {
        Row(
            modifier = Modifier.padding(16.dp),
            verticalAlignment = Alignment.CenterVertically
        ) {
            Icon(
                imageVector = Icons.Default.Security,
                contentDescription = null,
                tint = MaterialTheme.colorScheme.onSurfaceVariant
            )
            Spacer(modifier = Modifier.width(12.dp))
            Text(
                text = "Your biometric data is stored securely on your device and is never shared.",
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant
            )
        }
    }
}

class BiometricAuthViewModel : androidx.lifecycle.ViewModel() {
    var isAvailable by mutableStateOf(false)
        private set
    var biometricType by mutableStateOf(BiometricType.NONE)
        private set
    var availableBiometrics by mutableStateOf<List<BiometricType>>(emptyList())
        private set
    var selectedBiometric by mutableStateOf(BiometricType.NONE)
        private set
    var isLoading by mutableStateOf(false)
        private set
    var setupComplete by mutableStateOf(false)
        private set
    var authenticationSuccess by mutableStateOf(false)
        private set
    var error by mutableStateOf<String?>(null)
        private set

    var isSetup = false

    fun checkBiometricAvailability(context: android.content.Context) {
        val biometricManager = BiometricManager.from(context)

        when (biometricManager.canAuthenticate(BiometricManager.Authenticators.BIOMETRIC_STRONG)) {
            BiometricManager.BIOMETRIC_SUCCESS -> {
                isAvailable = true

                // Determine available biometric types
                val availableTypes = mutableListOf<BiometricType>()

                // Check for specific biometric types
                when (biometricManager.canAuthenticate(BiometricManager.Authenticators.BIOMETRIC_STRONG)) {
                    BiometricManager.BIOMETRIC_SUCCESS -> {
                        // Default to fingerprint as it's most common
                        availableTypes.add(BiometricType.FINGERPRINT)
                        biometricType = BiometricType.FINGERPRINT
                        selectedBiometric = BiometricType.FINGERPRINT
                    }
                }

                availableBiometrics = availableTypes
                error = null
            }
            BiometricManager.BIOMETRIC_ERROR_NO_HARDWARE -> {
                isAvailable = false
                error = "Biometric authentication is not available"
            }
            BiometricManager.BIOMETRIC_ERROR_HW_UNAVAILABLE -> {
                isAvailable = false
                error = "Biometric hardware is currently unavailable"
            }
            BiometricManager.BIOMETRIC_ERROR_NONE_ENROLLED -> {
                isAvailable = false
                error = "No biometrics enrolled on this device"
            }
            else -> {
                isAvailable = false
                error = "Biometric authentication is not available"
            }
        }
    }

    fun selectBiometric(type: BiometricType) {
        selectedBiometric = type
        biometricType = type
    }

    fun authenticateWithBiometric(activity: FragmentActivity) {
        isLoading = true
        error = null

        val executor = ContextCompat.getMainExecutor(activity)
        val biometricPrompt = BiometricPrompt(activity, executor, object : BiometricPrompt.AuthenticationCallback() {
            override fun onAuthenticationSucceeded(result: BiometricPrompt.AuthenticationResult) {
                super.onAuthenticationSucceeded(result)
                isLoading = false
                if (isSetup) {
                    setupComplete = true
                } else {
                    authenticationSuccess = true
                }
            }

            override fun onAuthenticationError(errorCode: Int, errString: CharSequence) {
                super.onAuthenticationError(errorCode, errString)
                isLoading = false
                error = when (errorCode) {
                    BiometricPrompt.ERROR_USER_CANCELED -> "Authentication was cancelled"
                    BiometricPrompt.ERROR_LOCKOUT -> "Too many failed attempts. Try again later"
                    BiometricPrompt.ERROR_LOCKOUT_PERMANENT -> "Too many failed attempts. Biometric authentication disabled"
                    else -> errString.toString()
                }
            }

            override fun onAuthenticationFailed() {
                super.onAuthenticationFailed()
                // Don't set loading to false here as user can retry
                error = "Authentication failed. Please try again"
            }
        })

        val promptInfo = BiometricPrompt.PromptInfo.Builder()
            .setTitle(if (isSetup) "Setup Biometric Authentication" else "Biometric Authentication")
            .setSubtitle(if (isSetup) "Enable biometric authentication for your account" else "Authenticate to continue")
            .setNegativeButtonText("Cancel")
            .build()

        biometricPrompt.authenticate(promptInfo)
    }
}

@Preview(showBackground = true)
@Composable
fun BiometricAuthScreenPreview() {
    MaterialTheme {
        BiometricAuthScreen(isSetup = true)
    }
}

@Preview(showBackground = true)
@Composable
fun BiometricAuthUnavailablePreview() {
    MaterialTheme {
        BiometricAuthScreen(
            viewModel = BiometricAuthViewModel().apply {
                // Would need to simulate unavailable state
            }
        )
    }
}

File: snippets/screens/flutter/biometric_auth_screen/3/biometric_auth_screen.dart
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

enum BiometricType { fingerprint, faceId, voice, none }

class BiometricAuthScreen extends StatefulWidget {
  final bool isSetup;

  const BiometricAuthScreen({super.key, this.isSetup = false});

  @override
  State<BiometricAuthScreen> createState() => _BiometricAuthScreenState();
}

class _BiometricAuthScreenState extends State<BiometricAuthScreen>
    with TickerProviderStateMixin {
  late AnimationController _animationController;
  late Animation<double> _scaleAnimation;
  late Animation<double> _opacityAnimation;

  bool _isLoading = false;
  bool _isAvailable = false;
  List<BiometricType> _availableBiometrics = [];
  BiometricType _selectedBiometric = BiometricType.none;
  String? _error;
  bool _setupComplete = false;

  @override
  void initState() {
    super.initState();
    _initializeAnimations();
    _checkBiometricAvailability();
  }

  void _initializeAnimations() {
    _animationController = AnimationController(
      duration: const Duration(milliseconds: 1500),
      vsync: this,
    );

    _scaleAnimation = Tween<double>(
      begin: 1.0,
      end: 1.2,
    ).animate(CurvedAnimation(
      parent: _animationController,
      curve: Curves.easeInOut,
    ));

    _opacityAnimation = Tween<double>(
      begin: 0.3,
      end: 1.0,
    ).animate(CurvedAnimation(
      parent: _animationController,
      curve: Curves.easeInOut,
    ));

    _animationController.repeat(reverse: true);
  }

  @override
  void dispose() {
    _animationController.dispose();
    super.dispose();
  }

  Future<void> _checkBiometricAvailability() async {
    try {
      // Simulate checking available biometrics
      await Future.delayed(const Duration(milliseconds: 500));

      // Mock available biometrics - in real app, use local_auth package
      final availableBiometrics = <BiometricType>[
        BiometricType.fingerprint,
        BiometricType.faceId,
      ];

      setState(() {
        _isAvailable = availableBiometrics.isNotEmpty;
        _availableBiometrics = availableBiometrics;
        _selectedBiometric = availableBiometrics.first;
      });
    } catch (e) {
      setState(() {
        _isAvailable = false;
        _error = 'Unable to check biometric availability';
      });
    }
  }

  Future<void> _authenticateWithBiometric() async {
    setState(() {
      _isLoading = true;
      _error = null;
    });

    try {
      // Simulate biometric authentication
      await Future.delayed(const Duration(seconds: 2));

      if (widget.isSetup) {
        setState(() => _setupComplete = true);
      } else {
        if (mounted) {
          ScaffoldMessenger.of(context).showSnackBar(
            const SnackBar(content: Text('Biometric authentication successful!')),
          );
          Navigator.of(context).pushReplacementNamed('/home');
        }
      }
    } on PlatformException catch (e) {
      String errorMessage;
      switch (e.code) {
        case 'NotAvailable':
          errorMessage = 'Biometric authentication is not available';
          break;
        case 'NotEnrolled':
          errorMessage = 'No biometrics enrolled on this device';
          break;
        case 'LockedOut':
          errorMessage = 'Too many failed attempts. Try again later';
          break;
        case 'UserCancel':
          errorMessage = 'Authentication was cancelled';
          break;
        default:
          errorMessage = 'Authentication failed. Please try again';
      }
      setState(() => _error = errorMessage);
    } catch (e) {
      setState(() => _error = 'Authentication failed. Please try again');
    } finally {
      if (mounted) setState(() => _isLoading = false);
    }
  }

  Future<void> _skipBiometric() async {
    if (widget.isSetup) {
      Navigator.of(context).pushReplacementNamed('/home');
    } else {
      Navigator.of(context).pushReplacementNamed('/login');
    }
  }

  IconData _getBiometricIcon(BiometricType type) {
    switch (type) {
      case BiometricType.fingerprint:
        return Icons.fingerprint;
      case BiometricType.faceId:
        return Icons.face;
      case BiometricType.voice:
        return Icons.record_voice_over;
      case BiometricType.none:
        return Icons.security;
    }
  }

  String _getBiometricName(BiometricType type) {
    switch (type) {
      case BiometricType.fingerprint:
        return 'Fingerprint';
      case BiometricType.faceId:
        return 'Face ID';
      case BiometricType.voice:
        return 'Voice';
      case BiometricType.none:
        return 'Biometric';
    }
  }

  String _getBiometricDescription(BiometricType type) {
    switch (type) {
      case BiometricType.fingerprint:
        return 'Use your fingerprint to authenticate';
      case BiometricType.faceId:
        return 'Use your face to authenticate';
      case BiometricType.voice:
        return 'Use your voice to authenticate';
      case BiometricType.none:
        return 'Use biometric authentication';
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: widget.isSetup
          ? AppBar(
              title: const Text('Setup Biometric Auth'),
              backgroundColor: Colors.transparent,
              elevation: 0,
              foregroundColor: Theme.of(context).textTheme.bodyLarge?.color,
            )
          : null,
      body: SafeArea(
        child: Padding(
          padding: const EdgeInsets.all(24.0),
          child: !_isAvailable ? _buildUnavailableView()
              : (_setupComplete ? _buildSetupCompleteView() : _buildAuthView()),
        ),
      ),
    );
  }

  Widget _buildAuthView() {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      children: [
        if (!widget.isSetup) const SizedBox(height: 60),

        AnimatedBuilder(
          animation: _animationController,
          builder: (context, child) {
            return Transform.scale(
              scale: _scaleAnimation.value,
              child: Opacity(
                opacity: _opacityAnimation.value,
                child: Icon(
                  _getBiometricIcon(_selectedBiometric),
                  size: 120,
                  color: Theme.of(context).primaryColor,
                ),
              ),
            );
          },
        ),
        const SizedBox(height: 40),

        Text(
          widget.isSetup ? 'Setup ${_getBiometricName(_selectedBiometric)}' : 'Biometric Authentication',
          style: Theme.of(context).textTheme.headlineMedium?.copyWith(
            fontWeight: FontWeight.bold,
          ),
          textAlign: TextAlign.center,
        ),
        const SizedBox(height: 12),

        Text(
          widget.isSetup
              ? 'Enable ${_getBiometricName(_selectedBiometric).toLowerCase()} authentication for quick and secure access to your account.'
              : _getBiometricDescription(_selectedBiometric),
          style: Theme.of(context).textTheme.bodyLarge?.copyWith(
            color: Colors.grey[600],
          ),
          textAlign: TextAlign.center,
        ),

        const SizedBox(height: 40),

        if (_availableBiometrics.length > 1) ...[
          Container(
            padding: const EdgeInsets.all(16),
            decoration: BoxDecoration(
              color: Colors.grey[100],
              borderRadius: BorderRadius.circular(12),
              border: Border.all(color: Colors.grey[300]!),
            ),
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                Text(
                  'Available Authentication Methods:',
                  style: Theme.of(context).textTheme.titleMedium?.copyWith(
                    fontWeight: FontWeight.bold,
                  ),
                ),
                const SizedBox(height: 12),

                ..._availableBiometrics.map((biometric) {
                  return RadioListTile<BiometricType>(
                    value: biometric,
                    groupValue: _selectedBiometric,
                    onChanged: (value) {
                      setState(() => _selectedBiometric = value!);
                    },
                    title: Row(
                      children: [
                        Icon(_getBiometricIcon(biometric)),
                        const SizedBox(width: 12),
                        Text(_getBiometricName(biometric)),
                      ],
                    ),
                    contentPadding: EdgeInsets.zero,
                  );
                }),
              ],
            ),
          ),
          const SizedBox(height: 32),
        ],

        if (_error != null) ...[
          Container(
            padding: const EdgeInsets.all(16),
            decoration: BoxDecoration(
              color: Colors.red[50],
              borderRadius: BorderRadius.circular(12),
              border: Border.all(color: Colors.red[200]!),
            ),
            child: Row(
              children: [
                Icon(Icons.error_outline, color: Colors.red[700]),
                const SizedBox(width: 12),
                Expanded(
                  child: Text(
                    _error!,
                    style: TextStyle(color: Colors.red[700]),
                  ),
                ),
              ],
            ),
          ),
          const SizedBox(height: 24),
        ],

        Container(
          padding: const EdgeInsets.all(16),
          decoration: BoxDecoration(
            color: Colors.blue[50],
            borderRadius: BorderRadius.circular(12),
            border: Border.all(color: Colors.blue[200]!),
          ),
          child: Row(
            children: [
              Icon(Icons.security, color: Colors.blue[700]),
              const SizedBox(width: 12),
              Expanded(
                child: Text(
                  'Your biometric data is stored securely on your device and is never shared.',
                  style: TextStyle(
                    color: Colors.blue[700],
                    fontSize: 13,
                  ),
                ),
              ),
            ],
          ),
        ),

        const Spacer(),

        ElevatedButton(
          onPressed: _isLoading ? null : _authenticateWithBiometric,
          style: ElevatedButton.styleFrom(
            minimumSize: const Size(double.infinity, 56),
          ),
          child: _isLoading
              ? const SizedBox(
                  height: 20,
                  width: 20,
                  child: CircularProgressIndicator(strokeWidth: 2),
                )
              : Text(widget.isSetup ? 'Enable ${_getBiometricName(_selectedBiometric)}' : 'Authenticate'),
        ),

        const SizedBox(height: 16),

        if (!widget.isSetup) ...[
          OutlinedButton(
            onPressed: () => Navigator.of(context).pushReplacementNamed('/login'),
            style: OutlinedButton.styleFrom(
              minimumSize: const Size(double.infinity, 56),
            ),
            child: const Text('Use Password Instead'),
          ),
        ] else ...[
          TextButton(
            onPressed: _skipBiometric,
            child: const Text('Skip for Now'),
          ),
        ],

        const SizedBox(height: 20),
      ],
    );
  }

  Widget _buildUnavailableView() {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      children: [
        const SizedBox(height: 100),

        Icon(
          Icons.error_outline,
          size: 80,
          color: Colors.orange[600],
        ),
        const SizedBox(height: 32),

        Text(
          'Biometric Authentication Unavailable',
          style: Theme.of(context).textTheme.headlineMedium?.copyWith(
            fontWeight: FontWeight.bold,
          ),
          textAlign: TextAlign.center,
        ),
        const SizedBox(height: 12),

        Text(
          _error ?? 'Biometric authentication is not available on this device or no biometrics are enrolled.',
          style: Theme.of(context).textTheme.bodyLarge?.copyWith(
            color: Colors.grey[600],
          ),
          textAlign: TextAlign.center,
        ),

        const SizedBox(height: 40),

        Container(
          padding: const EdgeInsets.all(16),
          decoration: BoxDecoration(
            color: Colors.orange[50],
            borderRadius: BorderRadius.circular(12),
            border: Border.all(color: Colors.orange[200]!),
          ),
          child: Column(
            children: [
              Icon(Icons.info_outline, color: Colors.orange[700]),
              const SizedBox(height: 8),
              Text(
                'To use biometric authentication:',
                style: TextStyle(
                  color: Colors.orange[700],
                  fontWeight: FontWeight.bold,
                ),
              ),
              const SizedBox(height: 8),
              Text(
                '• Enable fingerprint or face unlock in your device settings\n• Make sure at least one biometric is enrolled\n• Restart the app after enabling biometrics',
                style: TextStyle(color: Colors.orange[700]),
                textAlign: TextAlign.left,
              ),
            ],
          ),
        ),

        const Spacer(),

        ElevatedButton(
          onPressed: () => _checkBiometricAvailability(),
          style: ElevatedButton.styleFrom(
            minimumSize: const Size(double.infinity, 56),
          ),
          child: const Text('Check Again'),
        ),

        const SizedBox(height: 16),

        OutlinedButton(
          onPressed: widget.isSetup ? _skipBiometric : () => Navigator.of(context).pushReplacementNamed('/login'),
          style: OutlinedButton.styleFrom(
            minimumSize: const Size(double.infinity, 56),
          ),
          child: Text(widget.isSetup ? 'Skip Setup' : 'Use Password Instead'),
        ),

        const SizedBox(height: 20),
      ],
    );
  }

  Widget _buildSetupCompleteView() {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      children: [
        const SizedBox(height: 100),

        Container(
          padding: const EdgeInsets.all(24),
          decoration: BoxDecoration(
            color: Colors.green[50],
            shape: BoxShape.circle,
          ),
          child: Icon(
            Icons.check_circle,
            size: 80,
            color: Colors.green[600],
          ),
        ),
        const SizedBox(height: 32),

        Text(
          'Biometric Authentication Enabled!',
          style: Theme.of(context).textTheme.headlineMedium?.copyWith(
            fontWeight: FontWeight.bold,
          ),
          textAlign: TextAlign.center,
        ),
        const SizedBox(height: 12),

        Text(
          'You can now use ${_getBiometricName(_selectedBiometric).toLowerCase()} to quickly and securely access your account.',
          style: Theme.of(context).textTheme.bodyLarge?.copyWith(
            color: Colors.grey[600],
          ),
          textAlign: TextAlign.center,
        ),

        const SizedBox(height: 40),

        Container(
          padding: const EdgeInsets.all(16),
          decoration: BoxDecoration(
            color: Colors.green[50],
            borderRadius: BorderRadius.circular(12),
            border: Border.all(color: Colors.green[200]!),
          ),
          child: Column(
            children: [
              Row(
                children: [
                  Icon(Icons.check_circle, color: Colors.green[700]),
                  const SizedBox(width: 12),
                  Expanded(
                    child: Text(
                      'Biometric Authentication Features:',
                      style: TextStyle(
                        color: Colors.green[700],
                        fontWeight: FontWeight.bold,
                      ),
                    ),
                  ),
                ],
              ),
              const SizedBox(height: 12),
              Text(
                '• Quick and secure login\n• No need to remember passwords\n• Your biometric data stays on your device\n• Can be disabled anytime in settings',
                style: TextStyle(color: Colors.green[700]),
                textAlign: TextAlign.left,
              ),
            ],
          ),
        ),

        const Spacer(),

        ElevatedButton(
          onPressed: () => Navigator.of(context).pushReplacementNamed('/home'),
          style: ElevatedButton.styleFrom(
            minimumSize: const Size(double.infinity, 56),
          ),
          child: const Text('Continue to App'),
        ),

        const SizedBox(height: 16),

        OutlinedButton(
          onPressed: () {
            // Test biometric authentication
            _authenticateWithBiometric();
          },
          style: OutlinedButton.styleFrom(
            minimumSize: const Size(double.infinity, 56),
          ),
          child: const Text('Test Biometric Login'),
        ),

        const SizedBox(height: 20),
      ],
    );
  }
}